README.md
# Balena Plant Saver

![](/img/hero.png)

This is a Raspberry Pi balenaCloud starter project to help you water your precious plants. See temperature, humidity, and soil moisture levels. Set watering thresholds and timings to automate the watering of your plants.

![](https://github.com/balena-io-playground/balena-plant-saver/blob/master/img/img01.jpg?raw=true)

Access your plant's dashboard from anywhere in the world using balenaCloud. Add new plant monitors quickly and easily using the same application code on different devices. Use your imagination (and some extra parts) and turn balenaPlant into a balenaGarden. :)

![](/img/setup.png)

For a more detailed walkthrough, check out our [official project guide](https://www.balena.io/blog/keep-your-plant-friends-happy-and-healthy-with-automated-watering-balenaplant/) on the balena blog.

## Hardware you'll need
This project uses the Pimoroni Automation HAT to make things a bit more user-friendly. Here's the list of materials:

* Raspberry Pi ZeroWH/3B/4 (that's what we've tested so far)
* A 32GB+ SD card
* Pimoroni Automation HAT (a discontinued Automation pHAT should work as well)
* DHT22 temperature and humidity sensor
* Capacitive soil moisture sensor
* Float switch
* 5V peristaltic pump (with water tube-- some vendors don't include the tubing-- *remember the tubing!!!*)

## Software you'll need
Here's the software that you'll need to get going:

* a free [balenaCloud](https://dashboard.balena-cloud.com/signup) account (first ten devices are fully featured and free)
* [balenaEtcher](https://www.balena.io/etcher/) to burn OS images to SD cards
* (optional) [balenaCLI](https://www.balena.io/docs/reference/balena-cli/) if you want to hack on this project, push code locally, etc.

## Set up the hardware

![](/img/wiring-balenaplant.png)

This wiring diagram covers a build that uses the Pimoroni Automation HAT. Wiring will differ if using other hardware.

### Solder on DHT22 sensor

![](/img/dht-22.png)

Connect the power and ground to the 3.3V pins. Connect the data pin to the `SCLK` pin, which is GPIO pin 11 (you'll see this referenced in the plant watering code).

### Add capacitive moisture sensor

![](/img/moisture-sensor.png)

As seen on this [balenaForum post](https://forums.balena.io/t/building-a-smart-houseplant-monitor-and-waterer/9170/21), it's helpful to paint or coat the exposed electronic components of this sensor with outdoor paint or nail polish. Power the soil moisture sensor using the 5V terminals on the Automation pHAT, and insert the data wire into the `ADC 1` terminal (since the sensor is giving us an analog signal to convert).

### Add the water pump

![](/img/water-pump.png)

Power the peristaltic pump using the 5V terminal on the Automation pHAT. Connect the pump's ground wire to the `NO` (Normally Open) terminal.

### Add the float switch

![](/img/float-switch.png)

Power the float switch using the 5V terminal and connect the other wire into the `INPUT 1` terminal.

### Close the loop

![](/img/pump-relay-close.png)

Close the circuit for the water pump by using a wire to connect the `COM` terminal to `GND`.

### Set everything up

After preparing the Automation pHAT (soldering terminals on, etc.), connect it to the Raspberry Pi via GPIO pins.

At this point, you can set your plant watering test up. Add the moisture sensor to the plant's soil (or lack thereof in the case of this bamboo). Add the pumping end of the water pump into your water source and the dousing end into the plant's soil. NOTE: It's not clear on some pumps as to which port does what-- once you test, mark or tape a line to help tell which is which.

#### Build or 3D print additional accessories

![](/img/img02.jpg?raw=true)

The original project creators built 3D-printed parts to hold various parts or to enclose the electronics for the project. Check out this [forum thread](https://forums.balena.io/t/building-a-smart-houseplant-monitor-and-waterer/9170/21) to learn more.

## Add and deploy the application

### Deploy with balena

Once you're logged into your balenaCloud account (remember: first ten devices are free and fully-featured), click the button below to automagically deploy the project to your account.

[![Deploy with balena](https://balena.io/deploy.svg)](https://dashboard.balena-cloud.com/deploy)

Name your application, select your device type, and click `Create and deploy`. You'll be taken to your dashboard and our builders will start creating your application in the background.

### Add a device

Once that step is complete, add a new device by clicking `Add device` and selecting your device type. If you plan on using Wi-Fi, add your credentials here. Otherwise, proceed with an Ethernet cable connected to your internet setup. Complete the modal and your OS will download.

### Flash the OS onto your SD card

Insert the SD card into your computer, boot up Etcher and flash the downloaded balenaOS to your card. Insert the card into your device and wait for it to power on and show up on your dashboard within your app. Once the device downloads all the software, the application should be ready to use.

## Automated watering

By default, balenaPlant checks the water level of the plant approximately every 15 minutes. This is a reasonable time for a few reasons:

* running the pump infrequently prevents the motor from burning out
* watering with long intervals between each session allows soil or growing material to absorb water for a more accurate reading

You can change the pump delay time by changing the `Device variable` (located on the left-side menu in balenaCloud). Set `pump_delay` to a higher number for a longer wait time (more pump cycle counts) and a lower one for a shorter wait between watering. If properly set, you'll never overwater because even if the elapsed delay goes by, the pump won't start if the moisure level meets your requirements.

Other device variables to change include:

| Device variable       | Definition                                 | Default value |
|-----------------------|--------------------------------------------|---------------|
| target_soil_moisture  | Sets target moisture percentage            | 60            |
| target_soil_threshold | Sets threshold of moisture past target     | 15            |
| pump_delay            | Adds time between how often pump code runs | 15            |

*NOTE: When you change a device variable,it will reset the device.*

## Experiment and explore

Once you test a few things here and there, try other kinds of plants, adjust the soil moisture targets, and take one chore off your to-do list.

### Build a fleet of balenaPlant devices

Why automate the watering of one plant when you could grow and nurture many more? What if you upgrade the pump? What if you add LED grow lights?? The possibilities are endless.

### Additional resources

Check out our [official project guide](https://www.balena.io/blog/keep-your-plant-friends-happy-and-healthy-with-automated-watering-balenaplant/) on the balena blog.

Check out this [livestream build](https://www.youtube.com/watch?v=r0ZR6COwdRo) (uses the balenaCLI method) or [visit our Forums](https://forums.balena.io/t/building-a-smart-houseplant-monitor-and-waterer/9170/21) for more information.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.2.8
## (2021-06-03)

* Update version and remove fleetcta [Andrew Nhem]

# v0.2.7
## (2021-06-03)

* Pin MQTT to version 1.6 for now [Phil Wilson]

# v0.2.6
## (2021-05-13)

* Fixing YML for open fleet [Andrew Nhem]

# v0.2.5
## (2021-05-13)

* Update balena.yml [Andrew Nhem]

# v0.2.4
## (2021-05-13)

* Making edit to yml to conform to app page convention. [Andrew Nhem]

# v0.2.3
## (2021-05-13)

* Edits to balena.yml to be able to add to Hub as open fleet [Andrew Nhem]

# v0.2.2
## (2021-02-01)

* Update balena.yml to new contract format [andrew]

# v0.2.1
## (2021-02-01)

* Fix typo, readme edits [andrew]

# v0.2.0
## (2021-01-29)

* minor tweaks [andrew]
* Add balena.yml, fix readme [andrew]
* Add balena.yml and readme fixes [andrew]
* Add balena.yml and readme edits [andrew]

# v0.1.0
## (2021-01-29)

* Add repo.yml [andrew]
* Add host networking for MQTT connection [Phil Wilson]
* tabbing fix [Phil Wilson]
* Block-ifying project [andrew]
* Adding updates to plant-saver project [andrewnhem]
* Added (temporary) reading from the temp/humidity sensor on the Sense-HAT board, and LED Matrix driver [Chris Crocker-White]
* Changed Grafana to proper provisioning [Chris Crocker-White]
* Set up a class and monitoring of inputs [Chris Crocker-White]
* Initial commit reading & storing moisture [Chris Crocker-White]


plantsaver/start.py
import time
from plantinha import PlantSaver


plantsaver = PlantSaver()
loop_count = 6 # update the tags on first iteration
pump = False
pump_count = 1
pump_on_count = 0

while True:
    plantsaver.tick()
    print("====================================================")
    print("Moisture {:.1f}%".format(plantsaver.moisture_level))
    print("Temperature: {:.1f}C".format(plantsaver.temperature))
    print("Humidity: {:.1f}%".format(plantsaver.humidity))
    print("Status: "+plantsaver.status)
    print("Water level: "+str(plantsaver.water_left))
    print("Pump Cont: "+str(pump_count))
    print("Status code:"+plantsaver.status)

    # Check if water level is too dry and that the pump wasn't on for the past 15 minutes
    if plantsaver.status_code == 1 and pump_count >= plantsaver.pump_delay * 6:
        print("Turning pump ON for 10 seconds.")
        plantsaver.pump_water(True)
        pump_count = 0
    else:
        if pump_on_count < 2:
            pump_on_count = pump_on_count + 1
        else:
            pump_on_count = 0
            plantsaver.pump_water(False)
            pump_count = pump_count + 1

    # Update device tags every minute
    if loop_count == 6:
        plantsaver.update_device_tags()
        loop_count = 0
    else:
        loop_count = loop_count + 1

    time.sleep(10)


plantsaver/hts221.py
# Modified/classified from https://raw.githubusercontent.com/ControlEverythingCommunity/HTS221/master/Python/HTS221.py
import smbus2
import time

class HTS221:
    temperature_units = 'C'

    data = None
    data_timestamp = 0

    def __init__(self, temperature_units = 'C'):
        if temperature_units != 'C':
            self.temperature_units = 'F'

        self.bus = smbus2.SMBus(1)

        # HTS221 address, 0x5F(95)
        # Select average configuration register, 0x10(16)
        #		0x1B(27)	Temperature average samples = 256, Humidity average samples = 512
        self.bus.write_byte_data(0x5F, 0x10, 0x1B)
        # HTS221 address, 0x5F(95)
        # Select control register1, 0x20(32)
        #		0x85(133)	Power ON, Continuous update, Data output rate = 1 Hz
        self.bus.write_byte_data(0x5F, 0x20, 0x85)

        time.sleep(0.5)

        # HTS221 address, 0x5F(95)
        # Read Calibration values from non-volatile memory of the device
        # Humidity Calibration values
        # Read data back from 0x30(48), 1 byte
        val = self.bus.read_byte_data(0x5F, 0x30)
        H0 = val / 2

        # Read data back from 0x31(49), 1 byte
        val = self.bus.read_byte_data(0x5F, 0x31)
        H1 = val /2

        # Read data back from 0x36(54), 2 bytes
        val0 = self.bus.read_byte_data(0x5F, 0x36)
        val1 = self.bus.read_byte_data(0x5F, 0x37)
        H2 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

        # Read data back from 0x3A(58), 2 bytes
        val0 = self.bus.read_byte_data(0x5F, 0x3A)
        val1 = self.bus.read_byte_data(0x5F, 0x3B)
        H3 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

        self.humidity_calibration = [H0, H1, H2, H3]

        # Temperature Calibration values
        # Read data back from 0x32(50), 1 byte
        T0 = self.bus.read_byte_data(0x5F, 0x32)
        T0 = (T0 & 0xFF)

        # Read data back from 0x32(51), 1 byte
        T1 = self.bus.read_byte_data(0x5F, 0x33)
        T1 = (T1 & 0xFF)

        # Read data back from 0x35(53), 1 byte
        raw = self.bus.read_byte_data(0x5F, 0x35)
        raw = (raw & 0x0F)

        # Convert the temperature Calibration values to 10-bits
        T0 = ((raw & 0x03) * 256) + T0
        T1 = ((raw & 0x0C) * 64) + T1

        # Read data back from 0x3C(60), 2 bytes
        val0 = self.bus.read_byte_data(0x5F, 0x3C)
        val1 = self.bus.read_byte_data(0x5F, 0x3D)
        T2 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

        # Read data back from 0x3E(62), 2 bytes
        val0 = self.bus.read_byte_data(0x5F, 0x3E)
        val1 = self.bus.read_byte_data(0x5F, 0x3F)
        T3 = ((val1 & 0xFF) * 256) + (val0 & 0xFF)

        self.temperature_calibration = [T0, T1, T2, T3]


    def read_data(self):
        if self.data_timestamp < (int(round(time.time() * 1000)) - 1000):
            self.data = self.bus.read_i2c_block_data(0x5F, 0x28 | 0x80, 4)
            self.data_timestamp = int(round(time.time() * 1000))

    def get_temperature(self):
        self.read_data()

        temp = (self.data[3] * 256) + self.data[2]
        if temp > 32767 :
        	temp -= 65536
        cTemp = ((self.temperature_calibration[1] - self.temperature_calibration[0]) / 8.0) * (temp - self.temperature_calibration[2]) / (self.temperature_calibration[3] - self.temperature_calibration[2]) + (self.temperature_calibration[0] / 8.0)
        fTemp = (cTemp * 1.8 ) + 32

        if self.temperature_units == 'F':
            return fTemp
        else:
            return cTemp


    def get_humidity(self):
        self.read_data()

        humidity = (self.data[1] * 256) + self.data[0]
        humidity = ((1.0 * self.humidity_calibration[1]) - (1.0 * self.humidity_calibration[0])) * (1.0 * humidity - 1.0 * self.humidity_calibration[2]) / (1.0 * self.humidity_calibration[3] - 1.0 * self.humidity_calibration[2]) + (1.0 * self.humidity_calibration[0])

        return humidity


plantsaver/plantsaver.py
import os
import ADS1115
import gpiozero
from influxdb import InfluxDBClient
from balena import Balena

class PlantSaver:
    # moisture sensor config
    max_moisture = 1230
    min_moisture = 2600
    target_soil_moisture = 50 # this will be a balenaCloud config var
    target_soil_threshold = 10 # no action will be taken if current moisture is target +/- this value

    # define input/output pins
    pump_control_pin = 16
    float_switch_pin = 26

    # i2c device address definitions
    adc_i2c_address = 0x48
    led_matrix = 0x46
    led_matrix_type = 'sense-hat'

    # influx db details
    influx_db_name = 'plant-data'
    influx_db_host = 'influxdb'

    def __init__(self):
        self.ads = ADS1115.ADS1115(self.adc_i2c_address)
        self.pump = gpiozero.DigitalOutputDevice(self.pump_control_pin)
        self.float_switch = gpiozero.DigitalInputDevice(self.float_switch_pin)

        self.influx_client = InfluxDBClient(self.influx_db_host, 8086, database=self.influx_db_name)
        self.influx_client.create_database(self.influx_db_name)

        # set up an instance of the SDK - used for updating device tags
        self.balena = Balena()
        self.balena.auth.login_with_token(os.environ['BALENA_API_KEY'])

        # some instance status vars
        self.status = 'Starting'
        self.status_code = 0
        self.moisture_level = None
        self.pumping = False

    # Update the device tags with the moisture level and the status on balenaCloud
    # This means that you'll be able to see the status of the plant from the dashboard
    def update_device_tags(self):
        self.balena.models.tag.device.set(os.environ['BALENA_DEVICE_UUID'], 'Status', str(self.status))
        moisture_string = str(self.moisture_level)+'%'
        self.balena.models.tag.device.set(os.environ['BALENA_DEVICE_UUID'], 'Moisture', moisture_string)

    # Store the current instance measurements within InfluxDB
    def write_measurements(self):

        measurements = [
            {
                'measurement': 'plant-data',
                'fields': {
                    'moisture': float(self.moisture_level),
                    'pumping': int(self.pumping),
                    'water_left': int(self.water_left),
                    'status': int(self.status_code)
                }
            }
        ]

        self.influx_client.write_points(measurements)

    # Take a reading from the moisture sensor and the float switch and update the vars with the current status
    def update_sensors(self):
        self.read_moisture()
        self.read_float_switch()

    # Take a reading from the moisture sensor,
    # then convert this into a percentage based on the defined minimum and maximum
    def read_moisture(self):
        value = self.ads.readADCSingleEnded()
        range = self.min_moisture - self.max_moisture
        self.moisture_level = round((100 - (((value - self.max_moisture)/range)*100)),1)

    # Take a reading from the float switch and update the vars
    def read_float_switch(self):
        self.water_left = bool(self.float_switch.value)

    # Generate a status string so we have something to show in the logs
    # We also generate a status code which is used in the front end UI
    def update_status(self):
        if self.moisture_level < self.target_soil_moisture-self.target_soil_threshold:
            status = 'Too dry'
            self.status_code = 1
        elif self.moisture_level > self.target_soil_moisture+self.target_soil_threshold:
            status = 'Too wet'
            self.status_code = 2
        else:
            status = 'OK'
            self.status_code = 3


        if self.pumping == True:
            status = status + ', pump running'

        if self.water_left == False:
            status = status + ', water low'

        self.status = status

    # Refresh the relevant things - designed to be run once every 10 seconds
    def tick(self):
        self.update_sensors()
        self.update_status()
        self.write_measurements()


plantsaver/plantinha.py
import time
import os
import automationhat
import Adafruit_DHT
from balena import Balena
import json
import paho.mqtt.client as mqtt

class PlantSaver:

    def __init__(self):

        self.client                 = mqtt.Client("1")

        # Variables
        self.dht_sensor             = Adafruit_DHT.DHT22
        self.dht_pin                = int(self.set_variable("dht_pin", 11))
        self.max_value              = float(self.set_variable("max_value", 2.77)) 
        self.min_value              = float(self.set_variable("min_value", 1.46)) 
        self.target_soil_moisture   = int(self.set_variable("target_soil_moisture", 60))
        self.target_soil_threshold  = int(self.set_variable("target_soil_threshold", 15))
        self.pump_delay             = int(self.set_variable("pump_delay", 15))

        # Initial status
        self.status         = 'Starting'
        self.status_code    = 0
        self.moisture_level = None
        self.pumping        = False
        self.temperature    = 0
        self.humidity       = 0

        # set up an instance of the SDK - used for updating device tags
        self.balena = Balena()
        self.balena.auth.login_with_token(os.environ['BALENA_API_KEY'])

    # Checks if there is an environment variable set, otherwise save the default value
    def set_variable(self, name, default_value):
        if name in os.environ:
            self.value = os.environ.get(name)
        else: 
            self.value = default_value
        return self.value
        
    def read_moisture(self):
        self.moisture_level= 100-(automationhat.analog.one.read()-self.min_value)/((self.max_value-self.min_value)/100)

    def read_temperature_humidity(self):
        self.humidity, self.temperature = Adafruit_DHT.read_retry(self.dht_sensor, self.dht_pin)
    
    def update_sensors(self):
        self.read_moisture()
        self.read_temperature_humidity()
        self.read_float_switch()

    # Take a reading from the float switch and update the vars
    def read_float_switch(self):
        self.water_left = not bool(automationhat.input.one.read())

    # Update the device tags with the moisture level and the status on balenaCloud
    # This means that you'll be able to see the status of the plant from the dashboard
    def update_device_tags(self):
        self.balena.models.tag.device.set(os.environ['BALENA_DEVICE_UUID'], 'Status', str(self.status))
        moisture_string = str(round(self.moisture_level,2))+'%'
        self.balena.models.tag.device.set(os.environ['BALENA_DEVICE_UUID'], 'Moisture', moisture_string)

    # Store the current instance measurements within InfluxDB
    def write_measurements(self):
        
        self.client.connect("localhost")

        measurements = [
            {
                'measurement': 'plant-data',
                'fields': {
                    'moisture': float(self.moisture_level),
                    'pumping': int(self.pumping),
                    'water_left': int(self.water_left),
                    'status': int(self.status_code),
                    'temperature': float(self.temperature),
                    'humidity': float(self.humidity)
                }
            }
        ]
        msgInfo = self.client.publish("sensors", json.dumps(measurements))
        if False == msgInfo.is_published():
            msgInfo.wait_for_publish()
        self.client.disconnect()

    # Generate a status string so we have something to show in the logs
    # We also generate a status code which is used in the front end UI
    def update_status(self):
        if self.moisture_level < self.target_soil_moisture-self.target_soil_threshold:
            status = 'Too dry'
            self.status_code = 1
        elif self.moisture_level > self.target_soil_moisture+self.target_soil_threshold:
            status = 'Too wet'
            self.status_code = 2
        else:
            status = 'OK'
            self.status_code = 3

        if self.pumping:
            status = status + ', pump on'
        else: 
            status = status + ', pump off'

        if not self.water_left:
            status = status + ', water low'
        else:
            status = status + ', water normal'

        self.status = status

    # Pump water
    def pump_water(self, action):
        if action == True:
            automationhat.relay.one.on()
            self.pumping = True
        else:
            automationhat.relay.one.off()
            self.pumping = False

    # Refresh the relevant things - designed to be run once every 10 seconds
    def tick(self):
        self.update_sensors()
        self.update_status()
        self.write_measurements()